scl/paloalto/panos.conf

```diff
@@ -0,0 +1,113 @@
+          prefix("`prefix`")
+          delimiters(',')
+        );
+      };
+    } 
+    elif (match('TRAFFIC', value('`prefix`type'))) {
+      parser {
+        csv-parser(
+          columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time","src_ip","dest_ip","src_translated_ip","dest_translated_ip","rule","src_user","dest_user","app","vsys","src_zone","dest_zone","src_interface","dest_interface","log_forwarding_profile","future_use3","session_id","repeat_count","src_port","dest_port","src_translated_port","dest_translated_port","session_flags","transport","action","bytes","bytes_out","bytes_in","packets","start_time","duration","http_category","future_use4","sequence_number","action_flags","src_location","dest_location","future_use5","packets_out","packets_in","session_end_reason","devicegroup_level1","devicegroup_level2","devicegroup_level3","devicegroup_level4","vsys_name","dvc_name")
+          prefix("`prefix`")
+          delimiters(',')
+        );
+      };
+    } 
+    elif (match('HIPWATCH', value('`prefix`type'))) {
+      parser {
+        csv-parser(
+          columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time","src_user","vsys","host_name","os","src_ip","hip_name","hip_count","hip_type","future_use3","future_use4","sequence_number","action_flags","devicegroup_level1","devicegroup_level2","devicegroup_level3","devicegroup_level4","vsys_name","dvc_name")
+          prefix("`prefix`")
+          delimiters(',')
+        );
+      };
+    } 
+    elif (match('CORRELATION', value('`prefix`type'))) {
+      parser {
+        csv-parser(
+          columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time","src_ip","src_user","vsys","category","severity","devicegroup_level1","devicegroup_level2","devicegroup_level3","devicegroup_level4","vsys_name","dvc_name","vsys_id","object","object_id","evidence")
+          prefix("`prefix`")
+          delimiters(',')
+        );
+      };
+    } 
+    elif (match('USERID', value('`prefix`type'))) {
+      parser {
+        csv-parser(
+          columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time","vsys","src_ip","source_name","event_id","repeat_count","timeout_threshold","src_port","dest_port","source","source_type","sequence_number","action_flags","devicegroup_level1","devicegroup_level2","devicegroup_level3","devicegroup_level4","vsys_name","dvc_name")
+          prefix("`prefix`")
+          delimiters(',')
+        );
+      };
+    };
+  };
+};
+application panos[syslog] {
+    filter { message(".*,paloalto.*"); };
```

Comment:
Maybe we can find something unique to use.

As Kokan mentioned before, the parser has 2 parts: a generic one (where type is found) and a category-specific one (SYSTEM, etc.).
Maybe we can use the generic one to check the parsed "type" field against known categories (SYSTEM, CONFIG, THREAT, etc.). If it matches, we know we have a Pan-OS log.

I'll think about it.

Reply:
This does not need to be perfect. the panos-parser() block would drop messages if they don't match the expected format (the big if/elif/... construct does not have an else block at the end, so if none of the branches match the log would be dropped).

In addition to this, I would probably add a flags(drop-invalid) option to the initial csv-parser() and move the rewrite rules toward the end (e.g. the setting of $PROGRAM and $MESSAGE), when we already proved the format was good.

Then I would simply not use a filter in the application adapter. What this would cause is that we would try to parse the incoming log as a panos specific log which would hopefully fail already at csv-parser, or in case it passes that, we would quickly exit when checking it against the recognized values (SYSTEM, CORRELATION, etc).

A good filter would be very useful, but that's only a performance improvement. It could be as easy as matching the first couple of characters with a regexp, e.g. message("1,[0-9]+/[0-9]+/[0-9]+ "); the pluses in that regexp could be changed to the exact number of digits, using the pcre syntax I can't recall right now (IIRC you can specify repetition count in braces).

With the filter in place and the panos-parser() dropping unrecognized log messages we should be fine.

If another device logs like panos and we recognize as panos, that can possibly be a problem, but let's wait for the conflict to arise.

Reply:
Thanks Bazsi for the feedback.

>  I would probably add a flags(drop-invalid) option to the initial csv-parser() 

I don't think you can, as that initial parser matches on a line where are more columns then configured, and we don't use the "greedy" flag, which would result in dropping the message at the first step.

> move the rewrite rules toward the end (e.g. the setting of $PROGRAM and $MESSAGE), when we already proved the format was good.

That was my initial idea too, but then realized the rewrite rule initializes the MESSAGE macro for the csv-parsers. It wouldn't be needed if we can solve the syslog() parsing step.
Basically, panos messages are csv fields added onto a semi-syslog message (we have a valid hostname, but after that the csv fields are starting.)

See example message from PR:
```
<14>Apr 14 16:54:18 paloalto.test.net 1,2020/04/14 16:54:18,unknown,CONFIG,0,0,2020/04/14 16:54:18,10.0.10.55,,set,admin,Web,Succeeded, deviceconfig system,127,0x0,0,0,0,0,,paloalto
```

Reply:
If you look at my branch superseding this one, I've added explicit filters to drop formats that do not confirm to the format. I chose to require fields up to "device_name" which is common in all of the formats and should never be zero. Newer versions extend the format with extra fields, those are also added but remained optional.